### PR TITLE
cmd/evm: adds ability to run individual state test file

### PR DESCRIPTION
**Pull request by holiman**
This PR adds ability to run a given general state test, as requested by @winsvega . 

~~Note This is WIP so far, I expect to have to modify it a bit to suit the needs by the test team.~~ 

Examples: 
```
#build/bin/evm  statetest  /home/martin/go/src/github.com/ethereum/go-ethereum/tests/testdata/GeneralStateTests/stCreateTest/CREATE_EmptyContract.json
Test {EIP158 0}
====================
PASS
====================
Test {Frontier 0}
====================
PASS
====================
Test {Homestead 0}
====================
PASS
====================
Test {Byzantium 0}
====================
PASS
====================
Test {Constantinople 0}
====================
FAIL : unsupported fork "Constantinople"
====================
Test {EIP150 0}
====================
PASS
====================
```
Most common flags can be used, e..g `--json` and `--debug`
```
#build/bin/evm  --json statetest  /home/martin/go/src/github.com/ethereum/go-ethereum/tests/testdata/GeneralStateTests/stCreateTest/CREATE_EmptyContract.json | head -n30
Test {Byzantium 0}
====================
{"pc":0,"op":90,"gas":"0x8d5b8","gasCost":"0x2","memory":"0x","memSize":0,"stack":[],"depth":1,"error":null,"opName":"GAS"}
{"pc":1,"op":96,"gas":"0x8d5b6","gasCost":"0x3","memory":"0x","memSize":0,"stack":["0x8d5b6"],"depth":1,"error":null,"opName":"PUSH1"}
{"pc":3,"op":85,"gas":"0x8d5b3","gasCost":"0x4e20","memory":"0x","memSize":0,"stack":["0x8d5b6","0x0"],"depth":1,"error":null,"opName":"SSTORE"}
{"pc":4,"op":96,"gas":"0x88793","gasCost":"0x3","memory":"0x","memSize":0,"stack":[],"depth":1,"error":null,"opName":"PUSH1"}
{"pc":6,"op":96,"gas":"0x88790","gasCost":"0x3","memory":"0x","memSize":0,"stack":["0x20"],"depth":1,"error":null,"opName":"PUSH1"}
{"pc":8,"op":96,"gas":"0x8878d","gasCost":"0x3","memory":"0x","memSize":0,"stack":["0x20","0x0"],"depth":1,"error":null,"opName":"PUSH1"}
{"pc":10,"op":240,"gas":"0x8878a","gasCost":"0x7d03","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":["0x20","0x0","0x0"],"depth":1,"error":null,"opName":"CREATE"}
{"pc":0,"op":0,"gas":"0x7ea5d","gasCost":"0x0","memory":"0x","memSize":0,"stack":[],"depth":2,"error":null,"opName":"STOP"}
{"pc":11,"op":96,"gas":"0x80a87","gasCost":"0x3","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":["0xf1ecf98489fa9ed60a664fc4998db699cfa39d40"],"depth":1,"error":null,"opName":"PUSH1"}
{"pc":13,"op":85,"gas":"0x80a84","gasCost":"0x4e20","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":["0xf1ecf98489fa9ed60a664fc4998db699cfa39d40","0x1"],"depth":1,"error":null,"opName":"SSTORE"}
{"pc":14,"op":90,"gas":"0x7bc64","gasCost":"0x2","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":[],"depth":1,"error":null,"opName":"GAS"}
{"pc":15,"op":96,"gas":"0x7bc62","gasCost":"0x3","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":["0x7bc62"],"depth":1,"error":null,"opName":"PUSH1"}
{"pc":17,"op":85,"gas":"0x7bc5f","gasCost":"0x4e20","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":["0x7bc62","0x64"],"depth":1,"error":null,"opName":"SSTORE"}
{"pc":18,"op":0,"gas":"0x76e3f","gasCost":"0x0","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":[],"depth":1,"error":null,"opName":"STOP"}
PASS
====================
Test {Constantinople 0}
====================
FAIL : unsupported fork "Constantinople"
====================
Test {EIP150 0}

```

And without json, but with `--debug`:
```
#build/bin/evm  --debug statetest  /home/martin/go/src/github.com/ethereum/go-ethereum/tests/testdata/GeneralStateTests/stCreateTest/CREATE_EmptyContract.json
Test {Homestead 0}
====================
PASS
#### TRACE ####
GAS             pc=00000000 gas=578998 cost=2

PUSH1           pc=00000001 gas=578995 cost=3
Stack:
00000000  000000000000000000000000000000000000000000000000000000000008d5b6

SSTORE          pc=00000003 gas=558995 cost=20000
Stack:
00000000  0000000000000000000000000000000000000000000000000000000000000000
00000001  000000000000000000000000000000000000000000000000000000000008d5b6
Storage:
0000000000000000000000000000000000000000000000000000000000000000: 000000000000000000000000000000000000000000000000000000000008d5b6

PUSH1           pc=00000004 gas=558992 cost=3
Storage:
0000000000000000000000000000000000000000000000000000000000000000: 000000000000000000000000000000000000000000000000000000000008d5b6

PUSH1           pc=00000006 gas=558989 cost=3
Stack:
00000000  0000000000000000000000000000000000000000000000000000000000000020
Storage:
0000000000000000000000000000000000000000000000000000000000000000: 000000000000000000000000000000000000000000000000000000000008d5b6
```

### CODE REVIEW

File: cmd/evm/staterunner.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		return err
+	}
+	cfg := vm.Config{
+		Tracer: tracer,
+		Debug:  ctx.GlobalBool(DebugFlag.Name) || ctx.GlobalBool(MachineFlag.Name),
+	}
+
+	for _, test := range tests {
+		for _, st := range test.Subtests() {
+			fmt.Printf("Test %v\n", st)
```
Comment (fjl) [inline cmd/evm/staterunner.go]: Maybe these lines should be JSON, too. We can decide about the format:

Emit test start lines and traces as separate objects:

```json
{"test": "name", "fork": "Byzantium", "tx": 0}
{"pc": 0, "op": ...}
```

Emit each test as a single object:

```json
{"test": "name", "fork": "Byzantium", "tx": 0, "log": [
{"pc": 0, "op": ...},
...
]}
```

Or even:

```json
[
{"test": "name", "fork": "Byzantium", "tx": 0, "log": [
...
]},
{"test": "name", "fork": "Byzantium", "tx": 0, "log": [
...
]}
]
```

Comment (holiman) [inline cmd/evm/staterunner.go]: Yes, makes sense. Right now the output is a bit mish-mash. I'll wait to see if @winsvega has any additional requests before making changes.

Comment (winsvega) [inline cmd/evm/staterunner.go]: Yes output as a json object would be better for parsing the result value.

**Comment by holiman** [comment]
This is now the format on `stdout`: 


	#build/bin/evm --json statetest  /home/martin/go/src/github.com/ethereum/go-ethereum/tests/testdata/GeneralStateTests/stCreateTest/CREATE_EmptyContract.json  2>/dev/null
	[{"name":"CREATE_EmptyContract","pass":true,"fork":"EIP158"},{"name":"CREATE_EmptyContract","pass":true,"fork":"Frontier"},{"name":"CREATE_EmptyContract","pass":true,"fork":"Homestead"},{"name":"CREATE_EmptyContract","pass":true,"fork":"Byzantium"},{"name":"CREATE_EmptyContract","pass":false,"fork":"Constantinople","error":"unsupported fork \"Constantinople\""},{"name":"CREATE_EmptyContract","pass":true,"fork":"EIP150"}]

Which looks like this, with some pretty-printing: 

	[~/go/src/github.com/ethereum/go-ethereum]
	#build/bin/evm --json statetest  /home/martin/go/src/github.com/ethereum/go-ethereum/tests/testdata/GeneralStateTests/stCreateTest/CREATE_EmptyContract.json  2>/dev/null| jq
	[
		{
			"name": "CREATE_EmptyContract",
			"pass": false,
			"fork": "Constantinople",
			"error": "unsupported fork \"Constantinople\""
		},
		{
			"name": "CREATE_EmptyContract",
			"pass": true,
			"fork": "EIP150"
		},
		{
			"name": "CREATE_EmptyContract",
			"pass": true,
			"fork": "EIP158"
		},
		{
			"name": "CREATE_EmptyContract",
			"pass": true,
			"fork": "Frontier"
		},
		{
			"name": "CREATE_EmptyContract",
			"pass": true,
			"fork": "Homestead"
		},
		{
			"name": "CREATE_EmptyContract",
			"pass": true,
			"fork": "Byzantium"
		}
	]

If you also look at the `stderr`, all executions will be printed, and then the "results-json" comes last: 


	#build/bin/evm --json statetest  /home/martin/go/src/github.com/ethereum/go-ethereum/tests/testdata/GeneralStateTests/stCreateTest/CREATE_EmptyContract.json 

	...
	{"pc":15,"op":96,"gas":"0x7bc62","gasCost":"0x3","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":["0x7bc62"],"depth":1,"error":null,"opName":"PUSH1"}
	{"pc":17,"op":85,"gas":"0x7bc5f","gasCost":"0x4e20","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":["0x7bc62","0x64"],"depth":1,"error":null,"opName":"SSTORE"}
	{"pc":18,"op":0,"gas":"0x76e3f","gasCost":"0x0","memory":"0x0000000000000000000000000000000000000000000000000000000000000000","memSize":32,"stack":[],"depth":1,"error":null,"opName":"STOP"}
	[{"name":"CREATE_EmptyContract","pass":false,"fork":"Constantinople","error":"unsupported fork \"Constantinople\""},{"name":"CREATE_EmptyContract","pass":true,"fork":"EIP150"},{"name":"CREATE_EmptyContract","pass":true,"fork":"EIP158"},{"name":"CREATE_EmptyContract","pass":true,"fork":"Frontier"},{"name":"CREATE_EmptyContract","pass":true,"fork":"Homestead"},{"name":"CREATE_EmptyContract","pass":true,"fork":"Byzantium"}]


And the same goes for `debug` - it shows only on stderr: 

	00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
	00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
	Storage:
	0000000000000000000000000000000000000000000000000000000000000000: 000000000000000000000000000000000000000000000000000000000008d5b6
	0000000000000000000000000000000000000000000000000000000000000001: 000000000000000000000000f1ecf98489fa9ed60a664fc4998db699cfa39d40
	0000000000000000000000000000000000000000000000000000000000000064: 000000000000000000000000000000000000000000000000000000000007bc62

	[{"name":"CREATE_EmptyContract","pass":true,"fork":"Byzantium"},{"name":"CREATE_EmptyContract","pass":false,"fork":"Constantinople","error":"unsupported fork \"Constantinople\""},{"name":"CREATE_EmptyContract","pass":true,"fork":"EIP150"},{"name":"CREATE_EmptyContract","pass":true,"fork":"EIP158"},{"name":"CREATE_EmptyContract","pass":true,"fork":"Frontier"},{"name":"CREATE_EmptyContract","pass":true,"fork":"Homestead"}]


So, afaik, it's valid json on `stdout`, but it's not optimal for debugging. 
@winsvega PTAL

**Comment by winsvega** [comment]
need to add logs support. post state now has field logs : hash
https://github.com/ethereum/tests/blob/develop/GeneralStateTests/stExample/add11.json#L25

no need to check this field you could just skip it. but evm fails to parse json with logs.

**Comment by holiman** [comment]
Rebased on master

**Comment by winsvega** [comment]
are there problems with merging ?

**Comment by holiman** [comment]
@winsvega there's some build error in AppVeyor, but I don't think it's related to this particular PR: 

```
?   	github.com/ethereum/go-ethereum/swarm/fuse	[no test files]
ok  	github.com/ethereum/go-ethereum/swarm/network	0.139s	coverage: 16.7% of statements
ok  	github.com/ethereum/go-ethereum/swarm/network/kademlia	2.925s	coverage: 78.0% of statements
?   	github.com/ethereum/go-ethereum/swarm/services/swap	[no test files]
ok  	github.com/ethereum/go-ethereum/swarm/services/swap/swap	0.026s	coverage: 65.3% of statements
ok  	github.com/ethereum/go-ethereum/swarm/storage	4.223s	coverage: 59.4% of statements
?   	github.com/ethereum/go-ethereum/swarm/testutil	[no test files]
runtime: out of memory: cannot allocate 65536-byte block (1610612736 in use)
fatal error: out of memory
runtime: out of memory: cannot allocate 65536-byte block (1610612736 in use)
fatal error: out of memory
runtime stack:
runtime.throw(0x7c61ec, 0xd)
	C:/go/src/runtime/panic.go:596 +0x7c
runtime.(*mcache).refill(0x350000, 0x3, 0x741be010)
	C:/go/src/runtime/mcache.go:124 +0xe3
runtime.(*mcache).nextFree.func1()
	C:/go/src/runtime/malloc.go:538 +0x28
runtime.systemstack(0x12895500)
	C:/go/src/runtime/asm_386.s:337 +0x5e
runtime.mstart()
	C:/go/src/runtime/proc.go:1132
goroutine 31721 [running]:
```
I'm  hoping that this PR gets merged for next release.

**Comment by holiman** [comment]
Re adding logs support - I'm hoping that comes automatically when I rebased on master.

**Comment by winsvega** [comment]
right. logs are working. I am testing on this PR now.

**Comment by holiman** [comment]
Rebased on master again

**Comment by holiman** [comment]
Nice @karalabe, looks good to me!

**Pull request merged**